src/storage/wal_replay.cpp

```diff
@@ -370,9 +370,69 @@ void WriteAheadLogDeserializer::ReplayDropTable() {
 	catalog.DropEntry(context, info);
 }
 
+void ReplayWithoutIndex(ClientContext &context, Catalog &catalog, AlterInfo &info, const bool only_deserialize) {
+	if (only_deserialize) {
+		return;
+	}
+	catalog.Alter(context, info);
+}
+
+void ReplayIndexData(AttachedDatabase &db, BinaryDeserializer &deserializer, IndexStorageInfo &info) {
+	D_ASSERT(info.IsValid() && !info.name.empty());
+
+	auto &storage_manager = db.GetStorageManager();
+	auto &single_file_sm = storage_manager.Cast<SingleFileStorageManager>();
+	auto &block_manager = single_file_sm.block_manager;
+	auto &buffer_manager = block_manager->buffer_manager;
+
+	deserializer.ReadList(103, "index_storage", [&](Deserializer::List &list, idx_t i) {
+		auto &data_info = info.allocator_infos[i];
+
+		// Read the data into buffer handles and convert them to blocks on disk.
+		for (idx_t j = 0; j < data_info.allocation_sizes.size(); j++) {
+
+			// Read the data into a buffer handle.
+			auto buffer_handle = buffer_manager.Allocate(MemoryTag::ART_INDEX, block_manager->GetBlockSize(), false);
```

Comment:
Can we pass the `deserialize_only` in here, and if that is true only deserialize instead of making these modifications to the block manager?

Reply:
Maybe we also need to add a test for the index WAL serialization, to make sure we are not increasing the file size unnecessarily (similar to the reclaim space tests where we e.g. create an index in the WAL, replay it, then drop it in a loop and verify our db size does not increase) - see `test/sql/storage/reclaim_space/reclaim_space_lists.test_slow`